File: siglume_api_sdk/testing/recorder.py
"""Lightweight VCR-like recorder for Siglume SDK tests."""
from __future__ import annotations

import base64
import json
import re
import time
from enum import Enum
from pathlib import Path
from typing import Any, Callable, Mapping
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

import httpx

CASSETTE_VERSION = 1
_SECRET_KEY_RE = re.compile(r"(api[_-]?key|secret|private[_-]?key|access[_-]?token|refresh[_-]?token)", re.IGNORECASE)
_PRIVKEY_RE = re.compile(r"0x[a-f0-9]{64}")
_TOKEN_RE = re.compile(r"(pypi|ghp|gho|ghu|ghs)-[A-Za-z0-9]+")
_BOUNDARY_RE = re.compile(r'boundary="?([^";]+)"?', re.IGNORECASE)
HeaderValue = str | list[str]


class RecordMode(str, Enum):
    RECORD = "record"
    REPLAY = "replay"
    AUTO = "auto"


def _append_header(result: dict[str, HeaderValue], key: str, value: str) -> None:
    existing = result.get(key)
    if existing is None:
        result[key] = value
    elif isinstance(existing, list):
        existing.append(value)
    else:
        result[key] = [existing, value]


def _normalize_headers(headers: Mapping[str, Any] | httpx.Headers | None) -> dict[str, HeaderValue]:
    if headers is None:
        return {}
    result: dict[str, HeaderValue] = {}
    if isinstance(headers, httpx.Headers):
        items = headers.multi_items()
    else:
        items = headers.items()
    for key, value in items:
        normalized_key = str(key).lower()
        if isinstance(value, list):
            for entry in value:
                _append_header(result, normalized_key, str(entry))
        else:
            _append_header(result, normalized_key, str(value))
    return result


def _redact_string(value: str) -> str:
    value = _PRIVKEY_RE.sub("<REDACTED_PRIVKEY>", value)
    value = _TOKEN_RE.sub("<REDACTED_TOKEN>", value)
    return value


def _redact_header_value(key: str, value: str) -> str:
    if key == "content-type" and "multipart/form-data" in value.lower():
        return _normalize_multipart_content_type(value)
    if key == "authorization":
        # Preserve the scheme token so cassettes stay readable, but redact
        # every credential regardless of scheme (Bearer / Basic / Digest /
        # custom tokens). Falling through to _redact_string only catches
        # values that match our narrow secret regexes, which would leave
        # plenty of credentials in the clear.
        stripped = value.strip()
        if not stripped:
            return "<REDACTED>"
        head, sep, _ = stripped.partition(" ")
        # If the value has no whitespace separator, there is no scheme to
        # preserve — the entire value IS the credential (e.g. a bare
        # GitHub PAT `ghp_...` or a hex-encoded API key). Returning
        # `{head} <REDACTED>` in that case would echo the secret back.
        if not sep:
            return "<REDACTED>"
        return f"{head} <REDACTED>"
    if key in {"cookie", "set-cookie"} or _SECRET_KEY_RE.search(key):
        redacted = _redact_string(value)
        return redacted if redacted != value else "<REDACTED>"
    return _redact_string(value)


def _redact_headers(headers: Mapping[str, Any] | httpx.Headers | None) -> dict[str, HeaderValue]:
    normalized = _normalize_headers(headers)
    result: dict[str, HeaderValue] = {}
    for key, value in normalized.items():
        if isinstance(value, list):
            result[key] = [_redact_header_value(key, item) for item in value]
        else:
            result[key] = _redact_header_value(key, value)
    return result


def _redact_url(url: str) -> str:
    parts = urlsplit(url)
    if not parts.query:
        return url
    redacted_query: list[tuple[str, str]] = []
    for key, value in parse_qsl(parts.query, keep_blank_values=True):
        if _SECRET_KEY_RE.search(key):
            next_value = _redact_string(value)
            redacted_query.append((key, next_value if next_value != value else "<REDACTED>"))
        else:
            redacted_query.append((key, _redact_string(value)))
    return urlunsplit((parts.scheme, parts.netloc, parts.path, urlencode(redacted_query), parts.fragment))


def _redact_body(value: Any, *, key_name: str | None = None) -> Any:
    if key_name and _SECRET_KEY_RE.search(key_name):
        if isinstance(value, str):
            redacted = _redact_string(value)
            return redacted if redacted != value else "<REDACTED>"
        return "<REDACTED>"
    if isinstance(value, Mapping):
        return {
            str(child_key): _redact_body(child_value, key_name=str(child_key))
            for child_key, child_value in value.items()
        }
    if isinstance(value, list):
        return [_redact_body(item) for item in value]
    if isinstance(value, str):
        return _redact_string(value)
    return value


def _normalize_multipart_bytes(content: bytes, content_type: str | None) -> bytes:
    if not content_type:
        return content
    match = _BOUNDARY_RE.search(content_type)
    if not match:
        return content
    boundary = match.group(1).encode("utf-8")
    return content.replace(boundary, b"<BOUNDARY>")


def _normalize_multipart_content_type(content_type: str | None) -> str:
    if not content_type:
        return "multipart/form-data; boundary=<BOUNDARY>"
    return _BOUNDARY_RE.sub("boundary=<BOUNDARY>", content_type)


def _parse_body_bytes(content: bytes | str | None, *, content_type: str | None = None) -> Any:
    if content in (None, b"", ""):
        return None
    if content_type and "multipart/form-data" in content_type.lower():
        raw_bytes = content if isinstance(content, bytes) else content.encode("utf-8")
        return {
            "content_type": _normalize_multipart_content_type(content_type),
            "encoding": "base64",
            "base64": base64.b64encode(_normalize_multipart_bytes(raw_bytes, content_type)).decode("ascii"),
        }
    if isinstance(content, bytes):
        text = content.decode("utf-8", errors="replace")
    else:
        text = content
    try:
        return json.loads(text)
    except ValueError:
        return text


def _serialize_request_body(request: httpx.Request) -> Any:
    try:
        body = request.content
    except Exception:
        try:
            body = request.read()
        except Exception:
            return None
    return _parse_body_bytes(body, content_type=request.headers.get("content-type"))


def _serialize_response_body(response: httpx.Response) -> Any:
    try:
        body = response.content
    except Exception:
        return None
    return _parse_body_bytes(body)


def _normalize_body_for_match(body: Any, ignore_body_fields: set[str]) -> Any:
    if isinstance(body, Mapping):
        return {
            key: _normalize_body_for_match(body[key], set())
            for key in sorted(body)
            if key not in ignore_body_fields
        }
    if isinstance(body, list):
        return [_normalize_body_for_match(item, set()) for item in body]
    return body


def _request_signature(request_payload: Mapping[str, Any], ignore_body_fields: set[str]) -> str:
    normalized = {
        "method": str(request_payload.get("method") or "").upper(),
        "url": str(request_payload.get("url") or ""),
        "body": _normalize_body_for_match(request_payload.get("body"), ignore_body_fields),
    }
    return json.dumps(normalized, ensure_ascii=False, separators=(",", ":"))


def _response_from_cassette(response_payload: Mapping[str, Any], request: httpx.Request) -> httpx.Response:
    headers: list[tuple[str, str]] = []
    for key, value in _normalize_headers(response_payload.get("headers")).items():
        if isinstance(value, list):
            headers.extend((str(key), str(item)) for item in value)
        else:
            headers.append((str(key), str(value)))
    body = response_payload.get("body")
    if isinstance(body, str):
        content = body.encode("utf-8")
    elif body is None:
        content = b""
    else:
        content = json.dumps(body, ensure_ascii=False).encode("utf-8")
        if not any(key.lower() == "content-type" for key, _ in headers):
            headers.append(("content-type", "application/json"))
    return httpx.Response(
        int(response_payload.get("status") or 200),
        headers=headers,
        content=content,
        request=request,
    )


class Recorder:
    """Record and replay HTTP interactions as JSON cassettes."""

    def __init__(
        self,
        cassette_path: str | Path,
        *,
        mode: RecordMode = RecordMode.AUTO,
        ignore_body_fields: list[str] | None = None,
    ) -> None:
        self.cassette_path = Path(cassette_path)
        self.mode = mode
        self.ignore_body_fields = set(ignore_body_fields or [])
        self._effective_mode = mode
        self._interactions: list[dict[str, Any]] = []
        self._cursor = 0
        self._original_client_request: Callable[..., httpx.Response] | None = None
        self._original_module_request: Callable[..., httpx.Response] | None = None

    def __enter__(self) -> "Recorder":
        self._effective_mode = self._resolve_mode()
        self._interactions = self._load_cassette() if self._effective_mode == RecordMode.REPLAY else []
        self._cursor = 0
        self._patch_httpx()
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self._restore_httpx()
        if exc_type is None and self._effective_mode == RecordMode.RECORD:
            self._write_cassette()

    def wrap(self, client: Any) -> Any:
        return client

    def _resolve_mode(self) -> RecordMode:
        if self.mode == RecordMode.AUTO:
            return RecordMode.REPLAY if self.cassette_path.exists() else RecordMode.RECORD
        return self.mode

    def _load_cassette(self) -> list[dict[str, Any]]:
        if not self.cassette_path.exists():
            raise AssertionError(f"Cassette not found for replay: {self.cassette_path}")
        payload = json.loads(self.cassette_path.read_text(encoding="utf-8"))
        if payload.get("version") != CASSETTE_VERSION:
            raise AssertionError(f"Unsupported cassette version in {self.cassette_path}")
        interactions = payload.get("interactions")
        if not isinstance(interactions, list):
            raise AssertionError(f"Invalid cassette interactions in {self.cassette_path}")
        return [dict(item) for item in interactions if isinstance(item, Mapping)]

    def _write_cassette(self) -> None:
        self.cassette_path.parent.mkdir(parents=True, exist_ok=True)
        payload = {"version": CASSETTE_VERSION, "interactions": self._interactions}
        self.cassette_path.write_text(
            json.dumps(payload, indent=2, sort_keys=True, ensure_ascii=False) + "\n",
            encoding="utf-8",
        )

    def _patch_httpx(self) -> None:
        # Patch both sync Client.request AND AsyncClient.request. httpx.request
        # (module-level) internally constructs a transient sync Client and
        # calls Client.request, so the sync patch catches both paths there.
        # AsyncClient takes an entirely separate code path — without patching
        # it, async callers (a common pattern in app adapters) hit the real
        # network in REPLAY mode, making cassette-based tests non-deterministic
        # and leaking external calls.
        self._original_client_request = httpx.Client.request
        self._original_async_client_request = httpx.AsyncClient.request
        self._original_module_request = None
        recorder = self
        original_client_request = self._original_client_request
        original_async_client_request = self._original_async_client_request

        def client_request_wrapper(client_self: httpx.Client, method: str, url: Any, *args: Any, **kwargs: Any) -> httpx.Response:
            request = client_self.build_request(
                method,
                url,
                content=kwargs.get("content"),
                data=kwargs.get("data"),
                files=kwargs.get("files"),
                json=kwargs.get("json"),
                params=kwargs.get("params"),
                headers=kwargs.get("headers"),
                cookies=kwargs.get("cookies"),
            )
            if recorder._effective_mode == RecordMode.REPLAY:
                return recorder._replay_request(request)
            started = time.perf_counter()
            response = original_client_request(client_self, method, url, *args, **kwargs)
            duration_ms = int(round((time.perf_counter() - started) * 1000))
            recorder._record_interaction(request, response, duration_ms)
            return response

        async def async_client_request_wrapper(client_self: httpx.AsyncClient, method: str, url: Any, *args: Any, **kwargs: Any) -> httpx.Response:
            request = client_self.build_request(
                method,
                url,
                content=kwargs.get("content"),
                data=kwargs.get("data"),
                files=kwargs.get("files"),
                json=kwargs.get("json"),
                params=kwargs.get("params"),
                headers=kwargs.get("headers"),
                cookies=kwargs.get("cookies"),
            )
            if recorder._effective_mode == RecordMode.REPLAY:
                return recorder._replay_request(request)
            started = time.perf_counter()
            response = await original_async_client_request(client_self, method, url, *args, **kwargs)
            duration_ms = int(round((time.perf_counter() - started) * 1000))
            recorder._record_interaction(request, response, duration_ms)
            return response

        httpx.Client.request = client_request_wrapper
        httpx.AsyncClient.request = async_client_request_wrapper

    def _restore_httpx(self) -> None:
        if self._original_client_request is not None:
            httpx.Client.request = self._original_client_request
        if getattr(self, "_original_async_client_request", None) is not None:
            httpx.AsyncClient.request = self._original_async_client_request

    def _record_interaction(self, request: httpx.Request, response: httpx.Response, duration_ms: int) -> None:
        self._interactions.append(
            {
                "request": {
                    "method": request.method,
                    "url": _redact_url(str(request.url)),
                    "headers": _redact_headers(request.headers),
                    "body": _redact_body(_serialize_request_body(request)),
                },
                "response": {
                    "status": response.status_code,
                    "headers": _redact_headers(response.headers),
                    "body": _redact_body(_serialize_response_body(response)),
                    "duration_ms": duration_ms,
                },
            }
        )

    def _replay_request(self, request: httpx.Request) -> httpx.Response:
        if self._cursor >= len(self._interactions):
            raise AssertionError(
                f"Replay attempted unexpected HTTP call {request.method} {request.url} after cassette was exhausted."
            )
        expected = self._interactions[self._cursor]
        actual_request = {
            "method": request.method,
            "url": _redact_url(str(request.url)),
            "body": _redact_body(_serialize_request_body(request)),
        }
        if _request_signature(expected["request"], self.ignore_body_fields) != _request_signature(actual_request, self.ignore_body_fields):
            raise AssertionError(
                "Replay request mismatch.\n"
                f"Expected: {_request_signature(expected['request'], self.ignore_body_fields)}\n"
                f"Actual:   {_request_signature(actual_request, self.ignore_body_fields)}"
            )
        self._cursor += 1
        return _response_from_cassette(expected["response"], request)

File: siglume-api-sdk-ts/src/testing/recorder.ts
type FetchLike = typeof fetch;
type CassetteHeaderValue = string | string[];
declare const Deno:
  | {
      stat(path: string): Promise<unknown>;
      readTextFile(path: string): Promise<string>;
      writeTextFile(path: string, content: string): Promise<void>;
      mkdir(path: string, options?: { recursive?: boolean }): Promise<void>;
    }
  | undefined;

const CASSETTE_VERSION = 1;
const SECRET_KEY_RE = /(api[_-]?key|secret|private[_-]?key|access[_-]?token|refresh[_-]?token)/i;
const HANDLE_URL_KEY_RE = /(checkout[_-]?url|portal[_-]?url)/i;
const PRIVKEY_RE = /0x[a-f0-9]{64}/g;
const TOKEN_RE = /(pypi|ghp|gho|ghu|ghs)-[A-Za-z0-9]+/g;
const BOUNDARY_RE = /boundary="?([^";]+)"?/i;

export enum RecordMode {
  RECORD = "record",
  REPLAY = "replay",
  AUTO = "auto",
}

export interface RecorderOptions {
  mode?: RecordMode;
  ignore_body_fields?: string[];
}

export interface CassetteInteraction {
  request: {
    method: string;
    url: string;
    headers: Record<string, CassetteHeaderValue>;
    body: unknown;
  };
  response: {
    status: number;
    headers: Record<string, CassetteHeaderValue>;
    body: unknown;
    duration_ms: number;
  };
}

export interface CassetteFile {
  version: number;
  interactions: CassetteInteraction[];
}

function redactString(value: string): string {
  return value.replace(PRIVKEY_RE, "<REDACTED_PRIVKEY>").replace(TOKEN_RE, "<REDACTED_TOKEN>");
}

function appendHeader(result: Record<string, CassetteHeaderValue>, key: string, value: string): void {
  const existing = result[key];
  if (existing === undefined) {
    result[key] = value;
    return;
  }
  if (Array.isArray(existing)) {
    existing.push(value);
    return;
  }
  result[key] = [existing, value];
}

function redactHeaderValue(key: string, value: string): string {
  if (key === "content-type" && value.toLowerCase().includes("multipart/form-data")) {
    return normalizeMultipartContentType(value);
  }
  if (key === "authorization") {
    // Preserve the scheme token so cassettes stay readable, but redact
    // every credential regardless of scheme (Bearer / Basic / Digest /
    // custom tokens). Falling through to redactString only catches values
    // that match our narrow secret regexes, which would leave plenty of
    // credentials in the clear.
    const stripped = value.trim();
    if (!stripped) {
      return "<REDACTED>";
    }
    // If the value has no whitespace separator, there is no scheme to
    // preserve — the entire value IS the credential (e.g. a bare
    // GitHub PAT `ghp_...` or a hex-encoded API key). Returning
    // `${head} <REDACTED>` in that case would echo the secret back.
    if (!/\s/.test(stripped)) {
      return "<REDACTED>";
    }
    const head = stripped.split(/\s+/)[0] ?? "";
    return head ? `${head} <REDACTED>` : "<REDACTED>";
  }
  if (key === "cookie" || key === "set-cookie" || SECRET_KEY_RE.test(key)) {
    const redacted = redactString(value);
    return redacted !== value ? redacted : "<REDACTED>";
  }
  return redactString(value);
}

function redactHeaders(headers: Headers): Record<string, CassetteHeaderValue> {
  const result: Record<string, CassetteHeaderValue> = {};
  headers.forEach((value, key) => {
    const normalizedKey = key.toLowerCase();
    appendHeader(result, normalizedKey, redactHeaderValue(normalizedKey, value));
  });
  const getSetCookie = Reflect.get(headers as object, "getSetCookie");
  if (typeof getSetCookie === "function") {
    const setCookies = (getSetCookie as () => string[]).call(headers)
      .map((value) => redactHeaderValue("set-cookie", value));
    if (setCookies.length === 1 && setCookies[0] !== undefined) {
      result["set-cookie"] = setCookies[0];
    } else if (setCookies.length > 1) {
      result["set-cookie"] = setCookies;
    }
  }
  return result;
}

function redactUrl(urlText: string): string {
  const url = new URL(urlText);
  const nextParams = new URLSearchParams();
  for (const [key, value] of Array.from(url.searchParams.entries())) {
    if (SECRET_KEY_RE.test(key) || HANDLE_URL_KEY_RE.test(key)) {
      const redacted = redactString(value);
      nextParams.append(key, redacted !== value ? redacted : "<REDACTED>");
    } else {
      nextParams.append(key, redactString(value));
    }
  }
  url.search = nextParams.toString();
  return url.toString();
}

function redactBody(value: unknown, keyName?: string): unknown {
  if (keyName && HANDLE_URL_KEY_RE.test(keyName)) {
    return "<REDACTED>";
  }
  if (keyName && SECRET_KEY_RE.test(keyName)) {
    if (typeof value === "string") {
      const redacted = redactString(value);
      return redacted !== value ? redacted : "<REDACTED>";
    }
    return "<REDACTED>";
  }
  if (Array.isArray(value)) {
    return value.map((item) => redactBody(item));
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([key, child]) => [key, redactBody(child, key)]),
    );
  }
  if (typeof value === "string") {
    return redactString(value);
  }
  return value;
}

function sortKeysDeep<T>(value: T): T {
  if (Array.isArray(value)) {
    return value.map((item) => sortKeysDeep(item)) as T;
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.keys(value as Record<string, unknown>)
        .sort()
        .map((key) => [key, sortKeysDeep((value as Record<string, unknown>)[key])]),
    ) as T;
  }
  return value;
}

function normalizeTopLevelBody(body: unknown, ignoreBodyFields: Set<string>): unknown {
  if (!body || typeof body !== "object") {
    return body;
  }
  if (Array.isArray(body)) {
    return body.map((item) => normalizeTopLevelBody(item, new Set<string>()));
  }
  const record = body as Record<string, unknown>;
  return Object.fromEntries(
    Object.keys(record)
      .filter((key) => !ignoreBodyFields.has(key))
      .sort()
      .map((key) => [key, normalizeTopLevelBody(record[key], new Set<string>())]),
  );
}

function requestSignature(request: { method: string; url: string; body: unknown }, ignoreBodyFields: Set<string>): string {
  return JSON.stringify({
    method: request.method.toUpperCase(),
    url: request.url,
    body: normalizeTopLevelBody(request.body, ignoreBodyFields),
  });
}

async function parseTextBody(text: string): Promise<unknown> {
  if (!text) {
    return null;
  }
  try {
    return JSON.parse(text) as unknown;
  } catch {
    return text;
  }
}

function normalizeMultipartText(text: string, contentType: string | null): string {
  if (!contentType) {
    return text;
  }
  const match = BOUNDARY_RE.exec(contentType);
  if (!match?.[1]) {
    return text;
  }
  return text.split(match[1]).join("<BOUNDARY>");
}

function replaceAllBytes(source: Uint8Array, search: Uint8Array, replacement: Uint8Array): Uint8Array {
  if (search.length === 0) {
    return source;
  }
  const parts: number[] = [];
  for (let index = 0; index < source.length;) {
    let matched = true;
    for (let offset = 0; offset < search.length; offset += 1) {
      if (source[index + offset] !== search[offset]) {
        matched = false;
        break;
      }
    }
    if (matched) {
      parts.push(...replacement);
      index += search.length;
      continue;
    }
    parts.push(source[index] as number);
    index += 1;
  }
  return Uint8Array.from(parts);
}

function normalizeMultipartBytes(bytes: Uint8Array, contentType: string | null): Uint8Array {
  if (!contentType) {
    return bytes;
  }
  const match = BOUNDARY_RE.exec(contentType);
  if (!match?.[1]) {
    return bytes;
  }
  return replaceAllBytes(bytes, new TextEncoder().encode(match[1]), new TextEncoder().encode("<BOUNDARY>"));
}

function normalizeMultipartContentType(contentType: string | null): string {
  if (!contentType) {
    return "multipart/form-data; boundary=<BOUNDARY>";
  }
  return contentType.replace(BOUNDARY_RE, "boundary=<BOUNDARY>");
}

function toBase64(bytes: Uint8Array): string {
  const bufferCtor = Reflect.get(globalThis as object, "Buffer") as { from(data: Uint8Array): { toString(encoding: string): string } } | undefined;
  if (bufferCtor?.from) {
    return bufferCtor.from(bytes).toString("base64");
  }
  if (typeof btoa === "function") {
    let binary = "";
    bytes.forEach((value) => {
      binary += String.fromCharCode(value);
    });
    return btoa(binary);
  }
  throw new Error("No base64 encoder available for multipart cassette recording.");
}

async function parseRequestBody(request: Request): Promise<unknown> {
  const contentType = request.headers.get("content-type");
  if (contentType?.toLowerCase().includes("multipart/form-data")) {
    const bytes = new Uint8Array(await request.clone().arrayBuffer());
    return {
      content_type: normalizeMultipartContentType(contentType),
      encoding: "base64",
      base64: toBase64(normalizeMultipartBytes(bytes, contentType)),
    };
  }
  const text = await request.clone().text();
  if (!text) {
    return null;
  }
  return parseTextBody(text);
}

function buildResponseFromCassette(interaction: CassetteInteraction, request: Request): Response {
  const headers = new Headers();
  Object.entries(interaction.response.headers).forEach(([key, value]) => {
    if (Array.isArray(value)) {
      value.forEach((entry) => headers.append(key, entry));
      return;
    }
    headers.append(key, value);
  });
  const body = interaction.response.body;
  const payload = typeof body === "string" ? body : body == null ? "" : JSON.stringify(body);
  if (body !== null && body !== undefined && typeof body !== "string" && !headers.has("content-type")) {
    headers.set("content-type", "application/json");
  }
  return new Response(payload, {
    status: interaction.response.status,
    headers,
  });
}

async function fileExists(path: string): Promise<boolean> {
  if (typeof Deno !== "undefined" && typeof Deno.stat === "function") {
    try {
      await Deno.stat(path);
      return true;
    } catch {
      return false;
    }
  }
  try {
    const fs = await import("node:fs/promises");
    await fs.access(path);
    return true;
  } catch {
    return false;
  }
}

async function readTextFile(path: string): Promise<string> {
  if (typeof Deno !== "undefined" && typeof Deno.readTextFile === "function") {
    return Deno.readTextFile(path);
  }
  const fs = await import("node:fs/promises");
  return fs.readFile(path, "utf8");
}

async function writeTextFile(path: string, content: string): Promise<void> {
  if (typeof Deno !== "undefined" && typeof Deno.writeTextFile === "function") {
    const lastSlash = Math.max(path.lastIndexOf("/"), path.lastIndexOf("\\"));
    const dir = lastSlash >= 0 ? path.slice(0, lastSlash) : "";
    if (dir) {
      await Deno.mkdir(dir, { recursive: true });
    }
    await Deno.writeTextFile(path, content);
    return;
  }
  const fs = await import("node:fs/promises");
  const pathModule = await import("node:path");
  await fs.mkdir(pathModule.dirname(path), { recursive: true });
  await fs.writeFile(path, content, "utf8");
}

export class Recorder {
  readonly cassettePath: string;
  readonly mode: RecordMode;
  readonly ignore_body_fields: string[];
  private effectiveMode: RecordMode | null = null;
  private interactions: CassetteInteraction[] = [];
  private cursor = 0;
  private wrappedClients: Array<{ client: object; fetchImpl: FetchLike }> = [];

  constructor(cassettePath: string, options: RecorderOptions = {}) {
    this.cassettePath = cassettePath;
    this.mode = options.mode ?? RecordMode.AUTO;
    this.ignore_body_fields = options.ignore_body_fields ?? [];
  }

  static async open(cassettePath: string, options: RecorderOptions = {}): Promise<Recorder> {
    const recorder = new Recorder(cassettePath, options);
    return recorder.start();
  }

  async start(): Promise<this> {
    this.effectiveMode = await this.resolveMode();
    this.interactions = this.effectiveMode === RecordMode.REPLAY ? await this.loadCassette() : [];
    this.cursor = 0;
    return this;
  }

  async close(): Promise<void> {
    for (const wrapped of this.wrappedClients.splice(0)) {
      Reflect.set(wrapped.client, "fetchImpl", wrapped.fetchImpl);
    }
    if (this.effectiveMode === RecordMode.RECORD) {
      const payload: CassetteFile = { version: CASSETTE_VERSION, interactions: this.interactions };
      await writeTextFile(this.cassettePath, `${JSON.stringify(sortKeysDeep(payload), null, 2)}\n`);
    }
  }

  wrap<T extends object>(client: T): T {
    if (!this.effectiveMode) {
      throw new Error("Recorder.wrap() requires start() first.");
    }
    const currentFetch = Reflect.get(client, "fetchImpl");
    if (typeof currentFetch !== "function") {
      throw new Error("Recorder.wrap() expects a SiglumeClient-like object with fetchImpl.");
    }
    this.wrappedClients.push({ client, fetchImpl: currentFetch as FetchLike });
    Reflect.set(client, "fetchImpl", this.createFetch(currentFetch as FetchLike));
    return client;
  }

  async withGlobalFetch<T>(fn: () => Promise<T> | T): Promise<T> {
    if (!this.effectiveMode) {
      throw new Error("Recorder.withGlobalFetch() requires start() first.");
    }
    const originalFetch = globalThis.fetch;
    if (typeof originalFetch !== "function") {
      throw new Error("Global fetch is not available in this runtime.");
    }
    const recorderFetch = this.createFetch(originalFetch);
    Reflect.set(globalThis as object, "fetch", recorderFetch);
    try {
      return await fn();
    } finally {
      Reflect.set(globalThis as object, "fetch", originalFetch);
    }
  }

  private async resolveMode(): Promise<RecordMode> {
    if (this.mode === RecordMode.AUTO) {
      return await fileExists(this.cassettePath) ? RecordMode.REPLAY : RecordMode.RECORD;
    }
    return this.mode;
  }

  private async loadCassette(): Promise<CassetteInteraction[]> {
    if (!(await fileExists(this.cassettePath))) {
      throw new Error(`Cassette not found for replay: ${this.cassettePath}`);
    }
    const payload = JSON.parse(await readTextFile(this.cassettePath)) as Partial<CassetteFile>;
    if (payload.version !== CASSETTE_VERSION || !Array.isArray(payload.interactions)) {
      throw new Error(`Invalid cassette format: ${this.cassettePath}`);
    }
    return payload.interactions;
  }

  private createFetch(fetchImpl: FetchLike): FetchLike {
    const ignoreBodyFields = new Set(this.ignore_body_fields);
    return (async (input: RequestInfo | URL, init?: RequestInit) => {
      const request = new Request(input, init);
      const requestBody = await parseRequestBody(request);
      const requestRecord = {
        method: request.method,
        url: redactUrl(request.url),
        headers: redactHeaders(request.headers),
        body: redactBody(requestBody),
      };

      if (this.effectiveMode === RecordMode.REPLAY) {
        if (this.cursor >= this.interactions.length) {
          throw new Error(`Replay attempted unexpected fetch ${request.method} ${request.url}`);
        }
        const interaction = this.interactions[this.cursor];
        if (!interaction) {
          throw new Error(`Replay interaction missing at index ${this.cursor}`);
        }
        if (requestSignature(interaction.request, ignoreBodyFields) !== requestSignature(requestRecord, ignoreBodyFields)) {
          throw new Error(
            `Replay request mismatch.\nExpected: ${requestSignature(interaction.request, ignoreBodyFields)}\nActual:   ${requestSignature(requestRecord, ignoreBodyFields)}`,
          );
        }
        this.cursor += 1;
        return buildResponseFromCassette(interaction, request);
      }

      const started = Date.now();
      const response = await fetchImpl(input, init);
      const responseText = response.status === 204 ? "" : await response.clone().text();
      const responseBody = await parseTextBody(responseText);
      this.interactions.push({
        request: requestRecord,
        response: {
          status: response.status,
          headers: redactHeaders(response.headers),
          body: redactBody(responseBody),
          duration_ms: Date.now() - started,
        },
      });
      return response;
    }) satisfies FetchLike;
  }
}
